src/util/ProjectIO.as

```diff
@@ -207,48 +207,113 @@ public class ProjectIO {
 		}
 	}
 
+	// Load all images in all costumes from their image data, then call whenDone.
 	public function decodeAllImages(objList:Array, whenDone:Function, fail:Function = null):void {
-		// Load all images in all costumes from their image data, then call whenDone.
-		function imageDecoded():void {
-			for each (var o:* in imageDict) {
-				if (o == 'loading...') return; // not yet finished loading
+		// This should be called on success or failure of each image
+		function imageDone():void {
+			if (--numImagesToDecode == 0) {
+				allImagesLoaded(objList, imageDict, whenDone, fail);
 			}
-			allImagesLoaded();
 		}
-		var error:Boolean = false;
-		function decodeError():void {
-			if (error) return;
-			error = true;
-			if (fail != null) fail();
-		}
-		function allImagesLoaded():void {
-			if (error) return;
-			for each (c in allCostumes) {
+
+		var numImagesToDecode:int = 1; // start at 1 to prevent early finish
+		var imageDict:Dictionary = new Dictionary(); // maps image data to BitmapData
+		for each (var obj:ScratchObj in objList) {
+			for each (var c:ScratchCostume in obj.costumes) {
 				if ((c.baseLayerData != null) && (c.baseLayerBitmap == null)) {
-					var img:* = imageDict[c.baseLayerData];
-					if (img is BitmapData) c.baseLayerBitmap = img;
-					if (img is SVGElement) c.setSVGRoot(img, false);
+					++numImagesToDecode;
+					if (ScratchCostume.isSVGData(c.baseLayerData)) {
+						decodeSVG(c.baseLayerData, imageDict, imageDone);
+					}
+					else {
+						decodeImage(c.baseLayerData, imageDict, imageDone, imageDone);
+					}
+				}
+				if ((c.textLayerData != null) && (c.textLayerBitmap == null)) {
+					++numImagesToDecode;
+					decodeImage(c.textLayerData, imageDict, imageDone, imageDone);
 				}
-				if ((c.textLayerData != null) && (c.textLayerBitmap == null)) c.textLayerBitmap = imageDict[c.textLayerData];
 			}
-			for each (c in allCostumes) c.generateOrFindComposite(allCostumes);
-			whenDone();
 		}
+		// decrement the artificial 1 and also handle the case where no images are present.
+		imageDone();
+	}
 
-		var c:ScratchCostume;
-		var allCostumes:Array = [];
-		for each (var o:ScratchObj in objList) {
-			for each (c in o.costumes) allCostumes.push(c);
+	private function allImagesLoaded(objList:Array, imageDict:Dictionary, whenDone:Function, fail:Function):void {
+		var errorCostumes:Vector.<ScratchCostume> = new Vector.<ScratchCostume>(2);
+
+		function makeErrorImage(obj:ScratchObj, c:ScratchCostume):* {
+			// TODO: translation?
```

Comment:
Not really. Usually the "process" for translating anything in the editor is for me to add it to the `strings()` method, do the editor string export, and make it show up in Pootle.
The main reason I bailed on this was the degree of dynamic text, but that's because I forgot that I added [a way to add placeholders](https://github.com/LLK/scratch-flash/blob/3ed8b522be528b056a1b577630118bb42a95c227/src/uiwidgets/DialogBox.as#L108) to a translatable string. I think we only use that in one place so far (the "Project not saved" dialog in the online editor) but we could certainly use it here as well. I'll give that a shot tomorrow.